File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        set<char> ss;  // Use set<char> to store characters
        int L = 0, longest = 0;  // Initialize 'longest' to 0

        for (int R = 0; R < s.length(); R++) {  // R is the index, not the character itself
            while (ss.find(s[R]) != ss.end()) {  // If s[R] is already in the set
                ss.erase(s[L]);  // Remove the character at the 'L' pointer
                L++;  // Move the left pointer to the right
            }
            ss.insert(s[R]);  // Insert the current character into the set
            int w = R - L + 1;  // Calculate the window size
            longest = max(longest, w);  // Update the longest length
        }

        return longest;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        left = 0
        
        uniqueSet = set()
        maxLen = 0
        
        for right in range(len(s)):
            while(s[right] in uniqueSet):
                uniqueSet.remove(s[left])
                left += 1
            uniqueSet.add(s[right])
            maxLen = max(maxLen, right-left+1)
        return maxLen
        